### ISSUE TITLE
Active dataflow cancellation

### ISSUE THREAD

**Issue by benesch** [MaterializeInc/materialize#2392]
At the moment, cancellation does not actually prevent any computation from occurring; it just allows the session that issued the query to issue another query instead.

Implementing cancellation is rather challenging, as it will require tight integration with timely. We'll need to be able to reach in and pinch off dataflow fragments when the cancellation request comes in.

Forked from #1537.

/cc @frankmcsherry @rjnn

**Comment by rjnn** [comment]
Tagging this tentatively as 0.4, but if it is really hard I'm accommodating of pushing this to 0.5. I do want to point out that right now the UX of cancellation really sucks, and it really is quite possible to accidentally throw in a SELECT query with a cross-join in it and go 'ah crap' with no possible way to aggressively shut things down.

**Comment by frankmcsherry** [comment]
Timely now supports dataflow cancelation, though it isn't as well exercised as it could be. The known cases are tested and don't cause exceptions, but there were a few surprises and there could be more.

**Comment by frankmcsherry** [comment]
One gotcha that worried me early on was that for anything that is not a `SELECT`, we need to carefully avoid reporting incorrect data that result from "closing" inputs, which in differential means "done". This isn't conceptually tricky: we know for which times the outputs are correct (whatever the input timestamps are at at the moment of termination) but we want to make sure that information prevents transcription of any further timestamps in outputs.

**Comment by rjnn** [comment]
We should prioritize this up, as its a bit of a UX fail at the moment to hit Ctrl+C repeatedly and not get any response.

**Comment by frankmcsherry** [comment]
Sanity check, but Ctrl-C and dataflow cancelation aren't tightly coupled are they? The form should certainly be able to return before we've torn down a dataflow, or am I very confused about what it needs to do?

**Comment by benesch** [comment]
That's correct. This is covered in #1537, from which this issue was forked. Ctrl+C will return immediately _if_ Materialize is still interactive. If we've lost interactivity, we've got an operator that needs a fuel tank, and we'll need to track those down one by one.

Put another way, **if Ctrl+C hangs, file an issue with the specific query that hung.** We may also need the other materialized views that were installed on the system at the time—one of those views could be responsible for the loss of interactivity too.

**Comment by benesch** [comment]
The reason this isn't Ctrl+C specific is because if Ctrl+C is hanging, it means trivial queries like `SELECT * FROM view LIMIT 1` are also hanging.

**Comment by ggevay** [comment]
@teskje, if this turns out to not be possible to solve in the short term, could we at least warn the user when we didn't drop her dataflow? Because the user probably won't expect this omission at all, and then just get very confused about what is taking up all her resources. (Plus a hint that she should clean up by dropping the cluster would also be good.)

**Comment by ggevay** [comment]
In the meantime, https://github.com/MaterializeInc/materialize/pull/18718 solved this for the [specific case of WMR queries](https://github.com/MaterializeInc/materialize/issues/16800).
